[tado] Channels added: Battery Low Alarm, A/C Power, Open Window Detected

andrewfg opened the pull request:
I have modified this binding to add a Channel to check the Battery Low Alarm state for battery powered devices

andrewfg commented:
Please don’t merge this one yet. I woke up in the middle of last night with the realisation that it still needs some more work on it. It should be ready later today.

andrewfg commented:
Ok. My latest commit "TadoBatteryChecker instance moved from Tado Zone to Tado Home" has resolved all the open issues. (And it now handles the most unlikely case that there may be multiple Tado Home instances in an openHAB home). 

It is ready to merge, so I have removed the work in progress tag.

wborn commented:
Seems like all Jenkins agents are now offline because of memory issues @kaikreuzer. They ran out of swap space because there were again many zombie processes on those agents.

Are you also using Docker to run those Jenkins agents? I saw Java ran on PID 1. When running Java in a container it's also requried to use a sub reaper to clean up zombie processes. We use [tini](https://github.com/krallin/tini#why-tini) for that in the openHAB Docker container.

andrewfg commented:
I took advantage of the additional gained time to a) add a Channel for Air-Conditioner Power (Off, On).

And also I updated the ReadMe file to add descriptions for these newly added batteryLowAlarm and acPower Channels.

I am assuming that the mentioned Jenkins issues will be resolved by @kaikreuzer or @wborn since I have no idea what you guys are talking about :)

andrewfg commented:
In the meantime it has also passed the Travis and Jenkins tests. :)

Hilbrand commented on bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/BaseHomeThingHandler.java:
```diff
@@ -30,7 +30,9 @@ public BaseHomeThingHandler(Thing thing) {
-        return getHomeHandler().getHomeId();
+        // null pointer exception fixed by Andrew Fiddian-Green
+        TadoHomeHandler handler = getHomeHandler();
+        return handler != null ? handler.getHomeId() : new Long(0);
```
```suggestion
        return handler != null ? handler.getHomeId() : 0L;
```

andrewfg replied:
Ok, I will make this change. But for the avoidance of doubt, the method return type is "Long" rather than "long", so I want to be sure to return an actual Long, in order to avoid messing too deeply in the Original Author's code..

andrewfg replied:
I reverted to my original code below; the reason is that your proposal ( `0L` ) causes the binding to crash :(

`return handler != null ? handler.getHomeId() : new Long(0);`

Hilbrand commented on bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java:
```diff
@@ -0,0 +1,87 @@
+ * 
+ */
+public class TadoBatteryChecker {
+    private Logger logger = LoggerFactory.getLogger(TadoBatteryChecker.class);
```
```suggestion
    private final Logger logger = LoggerFactory.getLogger(TadoBatteryChecker.class);
```

andrewfg replied:
Ok. Done.

Hilbrand commented on bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java:
```diff
@@ -0,0 +1,87 @@
+                    zoneList = homeHandler.getApi().listZones(homeId);
+                } catch (IOException | ApiException e) {
+                    zoneList = null;
+                    logger.debug("Fetch (battery state) zone list exception {}", e);
```
The exception should not be in the string.
```suggestion
                    logger.debug("Fetch (battery state) zone list exception", e);
```

andrewfg replied:
Ok. Done.

Hilbrand commented on bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoZoneHandler.java:
```diff
@@ -64,7 +67,7 @@
     private ScheduledFuture<?> scheduledHvacChange;
     private GenericZoneCapabilities capabilities;
-
+    
```
```suggestion

```

andrewfg replied:
Ok. Done.

Hilbrand commented on bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoZoneHandler.java:
```diff
@@ -90,7 +93,9 @@ public OverlayTerminationCondition getDefaultTerminationCondition() throws IOExc
 
     public ZoneState getZoneState() throws IOException, ApiException {
-        return getApi().showZoneState(getHomeId(), getZoneId());
+        // null pointer exception fixed by Andrew Fiddian-Green
```
No need to put in specific comments like these.  You can make a remark in the commit message. If this has a issue number or other references I would mention that in the git commit message.

andrewfg replied:
Ok. Done.

wborn commented:
I'll find another channel to communicate about build problems instead of using the first comment box I run into! :smile:

andrewfg commented on bundles/org.openhab.binding.tado/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -45,6 +45,9 @@
 
 			<channel typeId="operationMode" id="operationMode"></channel>
+			
+			<channel typeId="batteryLowAlarm" id="batteryLowAlarm"></channel>
```
Ok, I will use the system type id. But it is only partially helpful since I need to add a binding specific description text to it (so the XML is not that much less).

andrewfg replied:
I decided finally NOT to use the system type id. The reason is that this channel must be read only, and the system type id is not so.

Hilbrand replied:
I created a channel using the system.low-battery typeId and it's readonly for me. Can you check if it really isn't readonly for you. The reason we prefer system channels is because having standard channels makes it less work to translate channels, and also would make it more easily to handle similar channels.
This channel definition is the only thing left. If we can handle this we can merge this pr.

andrewfg replied:
Ok. Reverted.

andrewfg commented on bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java:
```diff
@@ -0,0 +1,87 @@
+            // logger.debug("Fetching battery state for ZoneId {}", zoneId);
+            for (Zone zone : zoneList) {
+                if (zoneId == zone.getId()) {
+                    Boolean alarm = false;
```
Ok. Done.

andrewfg commented on bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java:
```diff
@@ -0,0 +1,87 @@
+                }
+            }
+        }
+        return OnOffType.from(false);
```
Ok. Done.

andrewfg commented on bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java:
```diff
@@ -0,0 +1,87 @@
+                    Boolean alarm = false;
+                    for (ControlDevice device : zone.getDevices()) {
+                        String batteryState = device.getBatteryState();
+                        alarm = alarm || (batteryState != null && !batteryState.equals("NORMAL"));
```
Some Tado devices are mains powered; in which case they return batteryState == null; however this is also NOT a low battery alarm for such devices, so I think its better to return OnOffType.OFF also for mains powered devices.

However I have made the other change that you suggested. Good catch!

andrewfg replied:
Ok, on reflection, I decided to follow your suggestion as follows..
- if any device has batteryState != "NORMAL" => State = ON
- else, if any device has batteryState != null => State = OFF
- else => State = UNDEF
-

andrewfg commented:
Whilst waiting for the merge, I added one more new channel -- namely switch that shows the status of the thermostat's "open window detection" function.

PS Jenkins seems to have gone away again ;(

andrewfg commented:
Oh, it looks like Jenkins passed again. Whoopee!

andrewfg commented:
Just to say that (IMHO) this one is now functioning properly, fully tested, stable, and ready to merge.

kaikreuzer commented on bundles/org.openhab.binding.tado/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -200,4 +204,27 @@
 		<description>ON if at home, OFF if away</description>
 	</channel-type>
 
+	<channel-type id="batteryLowAlarm">
```
Please don't introduce your own channel type for this but use `system.low-battery` instead (see https://www.openhab.org/docs/developer/bindings/thing-xml.html#system-state-channel-types).

kaikreuzer replied:
Oops, just saw that this had been discussed in https://github.com/openhab/openhab2-addons/pull/6057#discussion_r324465798 already.
If the system channel type is NOT read only, I would consider this a bug. I hence would still ask you to make use of it. (and the description should be fine, because if it is a channel of the zone, it is somewhat clear that the lowBattery is related to the zone).

andrewfg replied:
Ok Kai, I will revert to the system defined channel type. But can you please — therefore —raise a bug report that the system defined channel type must have the read only attribute tag?

andrewfg replied:
PS @kaikreuzer this binding also has temperature and humidity channels in the original code. So I suppose that if you want to force the usage of the system defined channel types, then logically I should also do that for those channels too. Or?? Personally I was trying to minimise my intervention in the original code to avoid breaking anything, but I can make those changes also, if you wish me too. What do you prefer? Is there a convention in OH about other persons making retrospective changes to other persons code?

PPS I suppose that I should therefore also go back to my other two bindings [siemensrds] and [neohub] and modify them to use system defined channel types for humidity and temperature? Or??

kaikreuzer replied:
> raise a bug report that the system defined channel type must have the read only attribute tag?

The system channel is defined as readOnly=true, see [here](https://github.com/openhab/openhab-core/blob/0f74348f5111fe94ef8069078bf57a426e1bca1b/bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/DefaultSystemChannelTypeProvider.java#L84). Can you really confirm that this is not the case? If so, feel free to enter a bug report, stating how you notice that this is not working as expected.

> So I suppose that if you want to force the usage of the system defined channel types, then logically I should also do that for those channels too. Or??

We do not ask anybody to adapt existing channels (as this might have some small impact on existing setups (e.g. label texts may change), but for newly introduced channels, we definitely wish that the system channel types are used.

andrewfg replied:
Ok. Thanks

andrewfg commented:
Ok, I reverted to the system channel-type, but I had to add a specific description text for the specific instance of this channel type.

Note: I think I made some mess up with the git commits, since I pushed some unrelated stuff that I had not intended to, and then I reverted those. I hope it is clean now, but please let me know.

martinvw commented:
> All looks good.
Github only showed the things-types.xml as being changed to me since the last time I viewed. That means nothing else changed.

Does that mean that it can be merged, or is there a reason you didn't do that yet :-)

Hilbrand commented:
No specific reason. Just Kai's blocking review. But that was addressed, as it was the same comment I made. So yes, it can be merged :tada:
